### PR TITLE
Vanishing Point Analysis

### PR DESCRIPTION
**Pull request by taokinbo**
This PR uses Canny edge detection and Hough Line to detect vanishing points within the photo. 

Currently our analysis:

- [ ] Detects and extends significant (nonvertical and nonhorizontal) lines

- [ ] Returns the coordinate of the point that has the minimized sum of distances to all the lines detected

TODO

- [ ] Tune parameters of Hough lines and Canny to filter out insignificant lines and find more accurate lines

- [ ] Change algorithm to detect intersections rather than minimized distances between coordinates and lines

@ryaanahmed @clarkce

### CODE REVIEW

File: backend/app/analysis/find_vanishing_point.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+                         index 1 = sum of distances from VP to all lines
+    TODO: If there is no vanishing point within the photo, return None or "offscreen"
+    TODO: Change method/metric used to find vanishing point to improve accuracy
+    """
+    step = 100
+    coords_to_dists_from_lines = {}
+    for i in range(0, x_pix, step):
+        for j in range(0, y_pix, step):
+            for l in lines:
+                x1, y1, x2, y2 = l[0]
```
Comment (ryaanahmed) [inline backend/app/analysis/find_vanishing_point.py:105]: please comment this math a bit!

File: backend/app/analysis/find_vanishing_point.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+    plt.subplot(121), plt.imshow(lines[0], cmap='gray')
+    plt.title('Edge Image'), plt.xticks([]), plt.yticks([])
+
+
+    filter_lines = []
+    # Filters out vertical and horizontal lines
+    for l in lines[1]:
+        try:
+            x1, y1, x2, y2 = l[0]
```
Comment (ryaanahmed) [inline backend/app/analysis/find_vanishing_point.py:39]: please comment the math a bit!

File: backend/app/analysis/find_vanishing_point.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
+"""
+
+vanishing_point_finder.py
```
Comment (ryaanahmed) [inline backend/app/analysis/find_vanishing_point.py:3]: keep in sync with name of file

**Review by ryaanahmed — CHANGES_REQUESTED** [review]
Looking really good... some things to fix before review:

- Bunch of Python linter failures (for the `x1` etc. not following naming conventions, the convention that fits our linter rules is  `x_1` etc.)

- The analyze function in `find_vanishing_point` is currently crashing the testing system on CI. I'm pretty sure it's because you're calling interactive show functions in opncv and matplotlib. These interactive widgets won't work on the CI server. Let's comment them out but leave them in the code (commented out) for debugging purposes.

**Comment by taokinbo** [comment]
Sorry about the mistakes Ryaan. My team is planning to fix them today so we can do a proper commit today or tomorrow. Thank you for clarifying the issues.

Best,
Tolu.

On Oct 13, 2020, at 1:30 PM, Ryaan Ahmed <notifications@github.com> wrote:

﻿

@ryaanahmed requested changes on this pull request.

Looking really good... some things to fix before review:

  *   Bunch of Python linter failures (for the x1 etc. not following naming conventions, the convention that fits our linter rules is x_1 etc.)

  *   The analyze function in find_vanishing_point is currently crashing the testing system on CI. I'm pretty sure it's because you're calling interactive show functions in opncv and matplotlib. These interactive widgets won't work on the CI server. Let's comment them out but leave them in the code (commented out) for debugging purposes.

________________________________

In backend/app/analysis/find_vanishing_point.py<https://github.com/dhmit/paris_1970/pull/65#discussion_r504133874>:

> +    finds the point that is closest to all lines on average.
+
+    :param lines: list of significant lines
+    :param x_pix: width of the image
+    :param y_pix: height of the image
+    :return: tuple with: index 0 = vanishing point coordinates
+                         index 1 = sum of distances from VP to all lines
+    TODO: If there is no vanishing point within the photo, return None or "offscreen"
+    TODO: Change method/metric used to find vanishing point to improve accuracy
+    """
+    step = 100
+    coords_to_dists_from_lines = {}
+    for i in range(0, x_pix, step):
+        for j in range(0, y_pix, step):
+            for l in lines:
+                x1, y1, x2, y2 = l[0]


please comment this math a bit!

________________________________

In backend/app/analysis/find_vanishing_point.py<https://github.com/dhmit/paris_1970/pull/65#discussion_r504134144>:

> +    image = photo.get_image_data()
+    # Convert image to grayscale
+    # (Changes image array shape from (height, width, 3) to (height, width))
+    # (Pixels (image[h][w]) will be a value from 0 to 255)
+    grayscale_image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
+    lines = auto_canny(grayscale_image)
+
+    plt.subplot(121), plt.imshow(lines[0], cmap='gray')
+    plt.title('Edge Image'), plt.xticks([]), plt.yticks([])
+
+
+    filter_lines = []
+    # Filters out vertical and horizontal lines
+    for l in lines[1]:
+        try:
+            x1, y1, x2, y2 = l[0]


please comment the math a bit!

________________________________

In backend/app/analysis/find_vanishing_point.py<https://github.com/dhmit/paris_1970/pull/65#discussion_r504134360>:

> @@ -0,0 +1,122 @@
+"""
+
+vanishing_point_finder.py


keep in sync with name of file

**Comment by ryaanahmed** [comment]
> Sorry about the mistakes Ryaan. My team is planning to fix them today so we can do a proper commit today or tomorrow. Thank you for clarifying the issues.

No apologies needed! This is what code review is for.

**Comment by taokinbo** [comment]
We have made changes and our analysis function now passes the linter and backend tests. 

@ryaanahmed @clarkce

**Comment by taokinbo** [comment]
We have made changes to the algorithm we are using to find the vanishing point and are closing the pull request until said changes are more finalized.

**Pull request closed without merging**